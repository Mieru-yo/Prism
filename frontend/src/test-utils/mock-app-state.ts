/*
 * SPDX-FileCopyrightText: 2025 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import * as useApplicationStateModule from '../hooks/common/use-application-state'
import type { ApplicationState } from '../redux'
import { initialState as initialStateDarkMode } from '../redux/dark-mode/initial-state'
import { initialState as initialStateEditorConfig } from '../redux/editor-config/initial-state'
import { initialState as initialStateNoteDetails } from '../redux/note-details/initial-state'
import { initialState as initialStateRealtimeStatus } from '../redux/realtime/initial-state'
import { initialState as initialStateRendererStatus } from '../redux/renderer-status/initial-state'
import { type DeepPartial, ProviderType } from '@hedgedoc/commons'
import { vi } from 'vitest'
import { initialState } from '../redux/user/initial-state'

let state: ApplicationState = {
  darkMode: initialStateDarkMode,
  editorConfig: initialStateEditorConfig,
  history: [], // Yes this allows no mocking and is therefore technically not correct, but the type is difficult to fix and we will remove it soon anyway.
  noteDetails: initialStateNoteDetails,
  printMode: false,
  realtimeStatus: initialStateRealtimeStatus,
  rendererStatus: initialStateRendererStatus,
  user: initialState
}

vi.mock('../redux/editor-config/methods', async (importOriginal) => ({
  ...(await importOriginal()),
  loadFromLocalStorage: vi.fn().mockReturnValue(undefined)
}))
vi.mock('../hooks/common/use-application-state', () => {
  return {
    useApplicationState: vi.fn(<TSelected>(selector: (state: ApplicationState) => TSelected) => {
      if (state === undefined) {
        throw new Error('application state not mocked')
      }
      return selector(state)
    })
  }
})

/**
 * Mocks the {@link ApplicationState} for a test.
 * When not overriden, it uses the initial state of the reducers.
 *
 * @param newState Overrides for the mocked state
 */
export const mockAppState = (newState?: DeepPartial<ApplicationState>) => {
  state = {
    darkMode: {
      ...initialStateDarkMode,
      ...newState?.darkMode
    },
    editorConfig: {
      ...initialStateEditorConfig,
      ...newState?.editorConfig
    },
    history: [], // Yes this allows no mocking and is therefore technically not correct, but the type is difficult to fix and we will remove it soon anyway.
    noteDetails: {
      ...initialStateNoteDetails,
      ...newState?.noteDetails
    },
    printMode: false,
    realtimeStatus: {
      ...initialStateRealtimeStatus,
      ...newState?.realtimeStatus
    },
    rendererStatus: {
      ...initialStateRendererStatus,
      ...newState?.rendererStatus
    },
    user: newState?.user
      ? {
          username: newState.user.username ?? '',
          email: newState.user.email ?? null,
          displayName: newState.user.displayName ?? '',
          photoUrl: newState.user.photoUrl ?? null,
          authProvider: newState.user.authProvider ?? ProviderType.LOCAL
        }
      : null
  }
}
