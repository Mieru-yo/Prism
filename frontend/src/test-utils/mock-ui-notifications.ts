/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import * as useUiNotificationsModule from '../components/notifications/ui-notification-boundary'
import { vi } from 'vitest'

vi.mock('../components/notifications/ui-notification-boundary')

/**
 * Mocks the {@link useUiNotifications} hook with stub functions for tests.
 */
export const mockUiNotifications = () => {
  vi.spyOn(useUiNotificationsModule, 'useUiNotifications').mockReturnValue({
    showErrorNotification: vi.fn(),
    dismissNotification: vi.fn(),
    dispatchUiNotification: vi.fn(),
    pruneNotification: vi.fn()
  })
}
