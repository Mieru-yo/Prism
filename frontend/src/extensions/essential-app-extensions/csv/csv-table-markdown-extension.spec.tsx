/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import type { CodeProps } from '../../../components/markdown-renderer/replace-components/code-block-component-replacer'
import { TestMarkdownRenderer } from '../../../components/markdown-renderer/test-utils/test-markdown-renderer'
import { mockI18n } from '../../../test-utils/mock-i18n'
import * as CsvTableModule from '../csv/csv-table'
import { CsvTableMarkdownExtension } from './csv-table-markdown-extension'
import { render } from '@testing-library/react'
import React from 'react'
import { describe, expect, it, vitest, beforeAll, afterAll } from 'vitest'
import { vi } from 'vitest'

vi.mock('../csv/csv-table')

describe('CSV Table Markdown Extension', () => {
  beforeAll(async () => {
    vitest.spyOn(CsvTableModule, 'CsvTable').mockImplementation((({ code }) => {
      return (
        <span>
          this is a mock for csv frame
          <code>{code}</code>
        </span>
      )
    }) as React.FC<CodeProps>)
    await mockI18n()
  })

  afterAll(() => {
    vitest.resetModules()
    vitest.restoreAllMocks()
  })

  it('renders a csv codeblock', () => {
    const view = render(
      <TestMarkdownRenderer extensions={[new CsvTableMarkdownExtension()]} content={'```csv\na;b;c\nd;e;f\n```'} />
    )
    expect(view.container).toMatchSnapshot()
  })
})
