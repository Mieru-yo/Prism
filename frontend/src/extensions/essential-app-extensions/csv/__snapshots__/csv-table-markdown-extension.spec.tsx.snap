// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSV Table Markdown Extension > renders a csv codeblock 1`] = `
<div>
  <pre>
    <code
      class="csv"
    >
      a;b;c
d;e;f

    </code>
  </pre>
  

</div>
`;
